### PR TITLE
Add hybridfs store type

### PR DESCRIPTION
**Pull request by danielmitterdorfer**
With this commit we introduce a new store type `hybridfs` that is a
hybrid between `mmapfs` and `niofs`. This store type chooses different
strategies to read Lucene files based on the read access pattern (random
or linear) in order to optimize performance.

This store type has been available in earlier versions of Elasticsearch
as `default_fs`. We have chosen a different name now in order to convey
the intent of the store type instead of tying it to the fact whether it
is the default choice.

### CODE REVIEW

**Comment by danielmitterdorfer** [comment]
I'm marking this PR as WIP for the time being putting it up for discussion and also to get CI expose early on.

File: server/src/test/java/org/elasticsearch/index/store/IndexStoreTests.java [hunk 1]
```diff
@@ -75,7 +79,7 @@ private void doTestStoreDirectory(Index index, Path tempDir, String typeSettingV
                     break;
                 case FS:
                     if (Constants.JRE_IS_64BIT && MMapDirectory.UNMAP_SUPPORTED) {
-                        assertTrue(directory.toString(), directory instanceof MMapDirectory);
+                        assertTrue(directory.toString(), directory instanceof FileSwitchDirectory);
```
Comment (jpountz) [inline server/src/test/java/org/elasticsearch/index/store/IndexStoreTests.java]: in the two above cases, maybe also check that directory.getPrimaryDir() is a MMapDirectory?

File: server/src/main/java/org/elasticsearch/index/store/FsDirectoryService.java [hunk 2]
```diff
@@ -30,17 +31,25 @@
...
 import java.nio.file.Path;
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.Set;
 
 public class FsDirectoryService extends DirectoryService {
+    /*
+     * We are mmapping norms, docvalues as well as term dictionaries, all other files are served through NIOFS
+     * this provides good random access performance while not creating unnecessary mmaps for files like stored
+     * fields etc.
```
[the remark this answers is not in the document]
Comment (danielmitterdorfer) [reply server/src/main/java/org/elasticsearch/index/store/FsDirectoryService.java]: I just restored the original comment but that makes sense to me. I'll update the comment accordingly.

File: server/src/test/java/org/elasticsearch/index/store/IndexStoreTests.java (hunk 1 above)
Comment (danielmitterdorfer) [reply server/src/test/java/org/elasticsearch/index/store/IndexStoreTests.java]: I added an additional assertion.

**Comment by danielmitterdorfer** [comment]
@elasticmachine retest this please

**Comment by danielmitterdorfer** [comment]
@elasticmachine retest this please

**Review by jpountz — APPROVED** [review]

**Review by jpountz — APPROVED** [review]

File: docs/reference/index-modules/store.asciidoc [hunk 3]
```diff
@@ -67,12 +67,22 @@ process equal to the size of the file being mapped. Before using this
...
+[[hybridfs]]`hybridfs`::
+
+The `hybridfs` type is a hybrid of `niofs` and `mmapfs`, which chooses the best
+file system type for each type of file based on the read access pattern.
+Currently only the Lucene term dictionary, norms and doc values files are
+memory mapped. All other files are opened using Lucene `NIOFSDirectory`.
+Similarly to `mmapfs` be sure you have allowed plenty of
+<<vm-max-map-count,virtual address space>>.
+
 [[allow-mmapfs]]
```
[the remark this answers is not in the document]
Comment (danielmitterdorfer) [reply docs/reference/index-modules/store.asciidoc]: @jpountz and me discussed this and we'll open a separate PR to deprecate `node.store.allow_mmapfs` and rename it to `node.store.allow_mmap`.

**Comment by danielmitterdorfer** [comment]
Thanks for the review! I'll open a separate PR for the backport to 6.x in which `hybridfs` will be included as an option but will not be the default choice.

**Comment by danielmitterdorfer** [comment]
I've raised a separate PR for the backport in https://github.com/elastic/elasticsearch/pull/37067.

File: docs/reference/index-modules/store.asciidoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (danielmitterdorfer) [reply docs/reference/index-modules/store.asciidoc]: I have raised https://github.com/elastic/elasticsearch/pull/37070.

**Comment by s1monw** [comment]
@danielmitterdorfer I think this implementation is broken with respect to managing pending deletes. I do have a fix for it. Just a heads-up for the backport. It will also fi #37111

**Comment by danielmitterdorfer** [comment]
@s1monw sorry to hear and thank you for the heads-up.

**Pull request merged**